packages/core/src/signals/Vector2CompoundSignalContext.ts

```diff
@@ -0,0 +1,247 @@
+import {InterpolationFunction, TimingFunction} from '../tweening';
+import {PossibleVector2, Vector2} from '../types';
+import {CompoundSignal, CompoundSignalContext} from './CompoundSignalContext';
+import {Signal} from './SignalContext';
+import {SignalExtensions, SignalGenerator, SignalValue} from './types';
+
+export type Vector2Signal<
+  TOwner = void,
+  TContext = Vector2CompoundSignalContext<TOwner>,
+> = CompoundSignal<PossibleVector2, Vector2, 'x' | 'y', TOwner, TContext> & {
+  edit(
+    callback: (current: Vector2) => SignalValue<PossibleVector2>,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+  scale(
+    value: number,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+  mul(
+    possibleVector: PossibleVector2,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+  div(
+    possibleVector: PossibleVector2,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+  add(
+    possibleVector: PossibleVector2,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+  sub(
+    possibleVector: PossibleVector2,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+  dot(
+    possibleVector: PossibleVector2,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+  cross(
+    possibleVector: PossibleVector2,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+  mod(
+    possibleVector: PossibleVector2,
+    duration?: number,
+    timingFunction?: TimingFunction,
+    interpolationFunction?: InterpolationFunction<Vector2>,
+  ): Vector2 | TOwner | SignalGenerator<PossibleVector2, Vector2>;
+};
```

Comment:
You cannot declare these methods using single signatures as this doesn't contain all the necessary information. Depending on the passed parameters, a different type will be returned. This requires method overloading.
For example, this code won't work right now:
```ts
rect().position.add(20).rotation(45);
```
We know that calling `add` with only the value will return the owner of the signal, allowing us to chain other operations. But the type system doesn't know this.

Instead, you can take inspiration from the `CodeSignal`

First, declare interfaces for your operations using overloading:
```ts
export interface Vector2Edit<TOwner> {
  (callback: (current: Vector2) => SignalValue<PossibleVector2>): TOwner;
  (
    callback: (current: Vector2) => SignalValue<PossibleVector2>,
    duration: number,
    timingFunction?: TimingFunction,
    interpolationFunction?: InterpolationFunction<Vector2>,
  ): SignalGenerator<PossibleVector2, Vector2>;
}

export interface Vector2Operation<TOwner> {
  (possibleVector: PossibleVector2): TOwner;
  (
    value: PossibleVector2,
    duration: number,
    timingFunction?: TimingFunction,
    interpolationFunction?: InterpolationFunction<Vector2>,
  ): SignalGenerator<PossibleVector2, Vector2>;
}
```
Then define the interface with all the helper methods:
```ts
export interface Vector2SignalHelpers<TOwner> {
  edit: Vector2Edit<TOwner>;
  mul: Vector2Operation<TOwner>;
  div: Vector2Operation<TOwner>;
  add: Vector2Operation<TOwner>;
  sub: Vector2Operation<TOwner>;
  dot: Vector2Operation<TOwner>;
  cross: Vector2Operation<TOwner>;
  mod: Vector2Operation<TOwner>;
}
```
Use it to both, declare the `Vector2Signal` type:
```ts
export type Vector2Signal<
  TOwner = void,
  TContext = Vector2CompoundSignalContext<TOwner>,
> = CompoundSignal<PossibleVector2, Vector2, 'x' | 'y', TOwner, TContext> &
  Vector2SignalHelpers<TOwner>;
```
And as an interface implemented by the context:
```ts
export class Vector2CompoundSignalContext<TOwner = void>
  extends CompoundSignalContext<PossibleVector2, Vector2, 'x' | 'y', TOwner>
  implements Vector2SignalHelpers<TOwner>
{
```
(This will make sure the context implements all the necessary methods)

Methods on the context should now also feature overloading:
```ts
class ... {
  public edit(
    callback: (current: Vector2) => SignalValue<PossibleVector2>,
  ): TOwner;
  public edit(
    callback: (current: Vector2) => SignalValue<PossibleVector2>,
    duration: number,
    timingFunction?: TimingFunction,
    interpolationFunction?: InterpolationFunction<Vector2>,
  ): SignalGenerator<PossibleVector2, Vector2>;
  public edit(
    callback: (current: Vector2) => SignalValue<PossibleVector2>,
    duration?: number,
    timingFunction?: TimingFunction,
    interpolationFunction?: InterpolationFunction<Vector2>,
  ): SignalGenerator<PossibleVector2, Vector2> | TOwner {
    const newValue = callback(this.get());
    return this.invoke(
      newValue,
      duration,
      timingFunction,
      interpolationFunction,
    ) as SignalGenerator<PossibleVector2, Vector2> | TOwner;
  }
}
```